mani_skill2/envs/tasks/dexterity/rotate_valve.py

```diff
@@ -0,0 +1,294 @@
+            env=self, robot_init_qpos_noise=self.robot_init_qpos_noise
+        )
+        self.table_scene.build()
+
+    def _load_articulations(self):
+        # Robel valve
+        if self.difficulty_level == 0:
+            # Only tri-valve
+            valve_angles_list = [(0, np.pi / 3 * 2, np.pi / 3 * 4)] * self.num_envs
+        elif self.difficulty_level == 1:
+            base_angles = [
+                np.arange(0, np.pi * 2, np.pi * 2 / 3),
+                np.arange(0, np.pi * 2, np.pi / 2),
+                np.arange(0, np.pi * 2, np.pi * 2 / 5),
+            ]
+            valve_angles_list = (
+                base_angles * int(self.num_envs // 3)
+                + base_angles[: int(self.num_envs % 3)]
+            )
+        elif self.difficulty_level == 2:
+            num_valve_head = self._main_rng.randint(3, 6, (self.num_envs,))
+            valve_angles_list = [
+                sample_valve_angles(num_head, self._main_rng)
+                for num_head in num_valve_head
+            ]
+        elif self.difficulty_level >= 3:
+            num_valve_head = self._main_rng.randint(3, 6, (self.num_envs,))
+            valve_angles_list = [
+                sample_valve_angles(num_head, self._main_rng)
+                for num_head in num_valve_head
+            ]
+        else:
+            raise ValueError(
+                f"Difficulty level must be a int within 0-4, but get {self.difficulty_level}"
+            )
+
+        valves: List[Articulation] = []
+        capsule_lens = []
+        for i, valve_angles in enumerate(valve_angles_list):
+            scene_mask = np.zeros(self.num_envs, dtype=bool)
+            scene_mask[i] = True
+            if self.difficulty_level < 3:
+                valve, capsule_len = build_robel_valve(
+                    self._scene,
+                    valve_angles=valve_angles,
+                    scene_mask=scene_mask,
+                    name=f"valve_station_{i}",
+                )
+            else:
+                scales = self._main_rng.randn(2) * 0.1 + 1
+                valve, capsule_len = build_robel_valve(
+                    self._scene,
+                    valve_angles=valve_angles,
+                    scene_mask=scene_mask,
+                    name=f"valve_station_{i}",
+                    radius_scale=scales[0],
+                    capsule_radius_scale=scales[1],
+                )
+            valves.append(valve)
+            capsule_lens.append(capsule_len)
+        self.valve = Articulation.merge(valves, "valve_station")
+        self.capsule_lens = torch.from_numpy(np.array(capsule_lens)).to(self.device)
+        self.valve_head_links = get_obj_by_name(self.valve.get_links(), "valve")
+
+    def _initialize_actors(self, env_idx: torch.Tensor):
+        with torch.device(self.device):
+            self.table_scene.initialize()
+
+        # Initialize task related information
+        with torch.device(self.device):
```

Comment:
small nit but if you want to reduce number of lines of code you can just do with torch.device once and remove the repeated ones